### PR TITLE
Added warnings for NaN propagation

### LINKED ISSUE #340: `nan` propagation in matrix multiplication
**Describe the bug**

Matrix multiplication does not propagate `nan` like numpy does.

**To Reproduce**

```python
import numpy as np
import sparse

A = np.eye(4)
A[2, 2] = np.nan

np.eye(4) @ A
# array([[ 1.,  0., nan,  0.],
#        [ 0.,  1., nan,  0.],
#        [ 0.,  0., nan,  0.],
#        [ 0.,  0., nan,  1.]])

(sparse.eye(4) @ sparse.COO(A)).todense()
# array([[ 1.,  0.,  0.,  0.],
#        [ 0.,  1.,  0.,  0.],
#        [ 0.,  0., nan,  0.],
#        [ 0.,  0.,  0.,  1.]])
```

**Expected behavior**

I would expect the same values from `numpy` and `sparse`. However, I see that `nan` breaks the expected sparsity pattern, which could make efficient implementation difficult. Maybe a warning or error would be appropriate?

**System**
 - OS and version: macOS 10.15.3
 - `sparse` version (`sparse.__version__`) '0.9.1'
 - NumPy version (`np.__version__`) '1.18.3'
 - Numba version (`numba.__version__`) '0.49.0'

**Additional context**

* Scipy sparse matrices have the same behavior: https://github.com/scipy/scipy/issues/7532
* I believe a similar topic was touched on in #1, for checking results of operations against the fill value.

### PR DESCRIPTION
**Pull request by H4R5H1T-007**
Fixes #340 with a warning for now.

### CODE REVIEW

**Comment by hameerabbasi** [comment]
You can run the benchmarks with `asv run`, [check the docs for `asv`](https://asv.readthedocs.io/en/stable/using.html)

File: sparse/_common.py [hunk 1]
```diff
@@ -33,6 +36,23 @@
...
+def nan_check(*args):
+    for i in literal_unroll(args):
+        if np.isnan(np.min(np.asarray(i))):
+            return True
+    return False
+
+
+def check_class_nan(test):
+    if isinstance(test, SparseArray):
+        return nan_check(test.data, test.fill_value)
```
Comment (hameerabbasi) [inline sparse/_common.py:49]:
```suggestion
        return nan_check(test.fill_value, test.data)
```

The check may be faster this way.

File: sparse/_common.py [hunk 2]
```diff
@@ -33,6 +36,23 @@
...
+@numba.njit
+def nan_check(*args):
+    for i in literal_unroll(args):
+        if np.isnan(np.min(np.asarray(i))):
+            return True
+    return False
+
+
+def check_class_nan(test):
+    if isinstance(test, SparseArray):
```
Comment (hameerabbasi) [inline sparse/_common.py:48]:
```suggestion
    if isinstance(test, (GCXS, COO)):
```

`DOK` doesn't have a `data` attribute.

**Comment by H4R5H1T-007** [comment]
```
All benchmarks:
       before           after         ratio
     [e036c254]       [318e6694]
     <master~2>       <master>  
      2.78±0.08ms      2.93±0.08ms     1.05  benchmark_coo.ElemwiseBroadcastingSuite.time_add
          361±2μs         376±10μs     1.04  benchmark_coo.ElemwiseBroadcastingSuite.time_mul
      3.42±0.03ms      3.67±0.07ms     1.07  benchmark_coo.ElemwiseSuite.time_add
         999±10μs      1.06±0.03ms     1.07  benchmark_coo.ElemwiseSuite.time_mul
          521±4μs          555±9μs     1.07  benchmark_coo.IndexingSuite.time_index_fancy
       26.1±0.1μs       27.3±0.4μs     1.05  benchmark_coo.IndexingSuite.time_index_scalar
          117±2μs          122±1μs     1.04  benchmark_coo.IndexingSuite.time_index_slice
         283±10μs          295±5μs     1.04  benchmark_coo.IndexingSuite.time_index_slice2
         299±10μs          307±2μs     1.03  benchmark_coo.IndexingSuite.time_index_slice3
         340±10μs         379±10μs    ~1.11  benchmark_coo.MatrixMultiplySuite.time_matmul
      4.72±0.04ms       5.02±0.1ms     1.06  benchmark_gcxs.ElemwiseBroadcastingSuite.time_add
      1.16±0.03ms      1.21±0.04ms     1.05  benchmark_gcxs.ElemwiseBroadcastingSuite.time_mul
       6.95±0.2ms      7.25±0.09ms     1.04  benchmark_gcxs.ElemwiseSuite.time_add
      3.14±0.04ms      3.33±0.07ms     1.06  benchmark_gcxs.ElemwiseSuite.time_mul
      1.57±0.02ms      1.68±0.01ms     1.07  benchmark_gcxs.IndexingSuite.time_index_fancy
          189±1μs          204±2μs     1.08  benchmark_gcxs.IndexingSuite.time_index_scalar
      1.56±0.03ms      1.65±0.02ms     1.06  benchmark_gcxs.IndexingSuite.time_index_slice
      1.03±0.01ms      1.10±0.03ms     1.07  benchmark_gcxs.IndexingSuite.time_index_slice2
         790±10μs         828±10μs     1.05  benchmark_gcxs.IndexingSuite.time_index_slice3
          342±3μs          369±6μs     1.08  benchmark_gcxs.MatrixMultiplySuite.time_matmul
          165±2ms          176±2ms     1.07  benchmark_tensordot.TensordotSuiteDenseSparse.time_dense
          165±2ms          176±4ms     1.06  benchmark_tensordot.TensordotSuiteDenseSparse.time_sparse
          134±2ms          146±6ms     1.09  benchmark_tensordot.TensordotSuiteSparseDense.time_dense
          134±2ms          145±4ms     1.09  benchmark_tensordot.TensordotSuiteSparseDense.time_sparse
         88.9±2ms         93.9±2ms     1.06  benchmark_tensordot.TensordotSuiteSparseSparse.time_dense
         88.3±2ms         92.8±1ms     1.05  benchmark_tensordot.TensordotSuiteSparseSparse.time_sparse
```

File: sparse/_common.py (hunk 2 above)
Comment (hameerabbasi) [reply sparse/_common.py:51]: This still needs applying.

File: sparse/_common.py (hunk 1 above)
Comment (hameerabbasi) [reply sparse/_common.py:52]: So does this one.

**Review by hameerabbasi — CHANGES_REQUESTED** [review]
A couple of small comments.

File: sparse/_common.py [hunk 3]
```diff
@@ -33,6 +36,26 @@
...
+        if np.isnan(np.min(np.asarray(i))):
+            return True
+    return False
+
+
+def check_class_nan(test):
+    from ._compressed import GCXS
+    from ._coo import COO
+
+    if isinstance(test, GCXS) or isinstance(test, COO):
```
Comment (hameerabbasi) [inline sparse/_common.py:51]:
```suggestion
    if isinstance(test, (GCXS, COO)):
```

This works just fine, but can be cleaned up a bit.

**Review by hameerabbasi — CHANGES_REQUESTED** [review]
One final thing.

File: sparse/_common.py (hunk 3 above)
Comment (H4R5H1T-007) [reply sparse/_common.py]: Sir this is final benchmark.
```
All benchmarks:

       before           after         ratio
     [e036c254]       [4053ce4b]
     <master~4>       <master>  
      2.78±0.08ms       2.85±0.1ms     1.03  benchmark_coo.ElemwiseBroadcastingSuite.time_add
          361±2μs         377±20μs     1.04  benchmark_coo.ElemwiseBroadcastingSuite.time_mul
      3.42±0.03ms       3.55±0.1ms     1.04  benchmark_coo.ElemwiseSuite.time_add
         999±10μs      1.02±0.04ms     1.02  benchmark_coo.ElemwiseSuite.time_mul
          521±4μs          551±4μs     1.06  benchmark_coo.IndexingSuite.time_index_fancy
       26.1±0.1μs       27.5±0.3μs     1.05  benchmark_coo.IndexingSuite.time_index_scalar
          117±2μs          120±3μs     1.02  benchmark_coo.IndexingSuite.time_index_slice
         283±10μs          277±9μs     0.98  benchmark_coo.IndexingSuite.time_index_slice2
         299±10μs          295±5μs     0.99  benchmark_coo.IndexingSuite.time_index_slice3
         340±10μs         348±10μs     1.02  benchmark_coo.MatrixMultiplySuite.time_matmul
      4.72±0.04ms       4.86±0.2ms     1.03  benchmark_gcxs.ElemwiseBroadcastingSuite.time_add
      1.16±0.03ms      1.19±0.04ms     1.03  benchmark_gcxs.ElemwiseBroadcastingSuite.time_mul
       6.95±0.2ms       6.88±0.2ms     0.99  benchmark_gcxs.ElemwiseSuite.time_add
      3.14±0.04ms       3.18±0.1ms     1.01  benchmark_gcxs.ElemwiseSuite.time_mul
+     1.57±0.02ms       1.77±0.3ms     1.13  benchmark_gcxs.IndexingSuite.time_index_fancy
          189±1μs          201±5μs     1.06  benchmark_gcxs.IndexingSuite.time_index_scalar
      1.56±0.03ms      1.63±0.01ms     1.04  benchmark_gcxs.IndexingSuite.time_index_slice
      1.03±0.01ms      1.07±0.02ms     1.04  benchmark_gcxs.IndexingSuite.time_index_slice2
         790±10μs         834±90μs     1.06  benchmark_gcxs.IndexingSuite.time_index_slice3
          342±3μs          378±6μs    ~1.11  benchmark_gcxs.MatrixMultiplySuite.time_matmul
          165±2ms        174±0.9ms     1.05  benchmark_tensordot.TensordotSuiteDenseSparse.time_dense
          165±2ms        176±0.8ms     1.06  benchmark_tensordot.TensordotSuiteDenseSparse.time_sparse
          134±2ms          144±2ms     1.07  benchmark_tensordot.TensordotSuiteSparseDense.time_dense
          134±2ms          146±2ms     1.09  benchmark_tensordot.TensordotSuiteSparseDense.time_sparse
         88.9±2ms         93.6±2ms     1.05  benchmark_tensordot.TensordotSuiteSparseSparse.time_dense
         88.3±2ms         92.2±2ms     1.04  benchmark_tensordot.TensordotSuiteSparseSparse.time_sparse

```

**Comment by H4R5H1T-007** [comment]
Sir this is final benchmark.
```
All benchmarks:

       before           after         ratio
     [e036c254]       [4053ce4b]
     <master~4>       <master>  
      2.78±0.08ms       2.85±0.1ms     1.03  benchmark_coo.ElemwiseBroadcastingSuite.time_add
          361±2μs         377±20μs     1.04  benchmark_coo.ElemwiseBroadcastingSuite.time_mul
      3.42±0.03ms       3.55±0.1ms     1.04  benchmark_coo.ElemwiseSuite.time_add
         999±10μs      1.02±0.04ms     1.02  benchmark_coo.ElemwiseSuite.time_mul
          521±4μs          551±4μs     1.06  benchmark_coo.IndexingSuite.time_index_fancy
       26.1±0.1μs       27.5±0.3μs     1.05  benchmark_coo.IndexingSuite.time_index_scalar
          117±2μs          120±3μs     1.02  benchmark_coo.IndexingSuite.time_index_slice
         283±10μs          277±9μs     0.98  benchmark_coo.IndexingSuite.time_index_slice2
         299±10μs          295±5μs     0.99  benchmark_coo.IndexingSuite.time_index_slice3
         340±10μs         348±10μs     1.02  benchmark_coo.MatrixMultiplySuite.time_matmul
      4.72±0.04ms       4.86±0.2ms     1.03  benchmark_gcxs.ElemwiseBroadcastingSuite.time_add
      1.16±0.03ms      1.19±0.04ms     1.03  benchmark_gcxs.ElemwiseBroadcastingSuite.time_mul
       6.95±0.2ms       6.88±0.2ms     0.99  benchmark_gcxs.ElemwiseSuite.time_add
      3.14±0.04ms       3.18±0.1ms     1.01  benchmark_gcxs.ElemwiseSuite.time_mul
+     1.57±0.02ms       1.77±0.3ms     1.13  benchmark_gcxs.IndexingSuite.time_index_fancy
          189±1μs          201±5μs     1.06  benchmark_gcxs.IndexingSuite.time_index_scalar
      1.56±0.03ms      1.63±0.01ms     1.04  benchmark_gcxs.IndexingSuite.time_index_slice
      1.03±0.01ms      1.07±0.02ms     1.04  benchmark_gcxs.IndexingSuite.time_index_slice2
         790±10μs         834±90μs     1.06  benchmark_gcxs.IndexingSuite.time_index_slice3
          342±3μs          378±6μs    ~1.11  benchmark_gcxs.MatrixMultiplySuite.time_matmul
          165±2ms        174±0.9ms     1.05  benchmark_tensordot.TensordotSuiteDenseSparse.time_dense
          165±2ms        176±0.8ms     1.06  benchmark_tensordot.TensordotSuiteDenseSparse.time_sparse
          134±2ms          144±2ms     1.07  benchmark_tensordot.TensordotSuiteSparseDense.time_dense
          134±2ms          146±2ms     1.09  benchmark_tensordot.TensordotSuiteSparseDense.time_sparse
         88.9±2ms         93.6±2ms     1.05  benchmark_tensordot.TensordotSuiteSparseSparse.time_dense
         88.3±2ms         92.2±2ms     1.04  benchmark_tensordot.TensordotSuiteSparseSparse.time_sparse

```

**Review by hameerabbasi — APPROVED** [review]
Waiting on CI.

**Pull request merged**